activerecord/lib/active_record/counter_cache.rb

```diff
@@ -46,15 +46,18 @@ def reset_counters(id, *counters, touch: nil)
           child_class  = has_many_association.klass
           reflection   = child_class._reflections.values.find { |e| e.belongs_to? && e.foreign_key.to_s == foreign_key && e.options[:counter_cache].present? }
           counter_name = reflection.counter_cache_column
+          old_count = object.send(counter_name)
+          new_count = object.send(counter_association).count(:all)
+          updates = {}
 
-          updates = { counter_name => object.send(counter_association).count(:all) }
+          updates[counter_name] = new_count unless new_count == old_count
 
           if touch
```

Comment:
We could also skip the `UPDATE` when `touch: true` by making this:

```ruby
if touch && updates.any?
  ...
end
```
This would be a change in behavior, but perhaps is a Good Idea:tm:?

Reply:
Interesting idea - do specs fail when you make either change?

Reply:
@oniofchaos Yes, skipping the `UPDATE` when `touch: true` (or `touch: :some_other_column`) does cause three tests to fail. But the counter cache itself didn't need to change in those three cases. This is a change in behavior, but not necessarily _wrong_.

```
$ bundle exec ruby -Itest test/cases/counter_cache_test.rb
Using sqlite3
Run options: --seed 30471

# Running:

.................F

Failure:
CounterCacheTest#test_reset_counters_with_touch:_:written_on [test/cases/counter_cache_test.rb:295]:
Expected 2018-09-28 14:30:13 UTC to be < 2018-09-28 14:30:13 UTC.


rails test test/cases/counter_cache_test.rb:294

............F

Failure:
CounterCacheTest#test_reset_counters_with_touch:_%i(_updated_at_written_on_) [test/cases/counter_cache_test.rb:332]:
Expected 2018-09-28 14:30:13 UTC to be < 2018-09-28 14:30:13 UTC.


rails test test/cases/counter_cache_test.rb:331

.F

Failure:
CounterCacheTest#test_reset_counters_with_touch:_true [test/cases/counter_cache_test.rb:258]:
Expected 2018-09-28 14:30:13 UTC to be < 2018-09-28 14:30:13 UTC.


rails test test/cases/counter_cache_test.rb:257

.........

Finished in 0.388417s, 108.1312 runs/s, 159.6223 assertions/s.
42 runs, 62 assertions, 3 failures, 0 errors, 0 skips
```

Reply:
I'll leave it up to a Rails maintainer to weigh in, but might be a good idea to make a separate PR to discuss actual breaking changes vs non-breaking changes like what's in here currently

Reply:
IFF this PR is merged, I'll open another to discuss the possibility of the behavior change.